File: Advent_of_code/Advent_of_code_05.py
"""For this task we will rearrange some boxes that look like this:

                    [L] [M]         [M]    
                [D] [R] [Z]         [C] [L]
                [C] [S] [T] [G]     [V] [M]
        [R]     [L] [Q] [B] [B]     [D] [F]
        [H] [B] [G] [D] [Q] [Z]     [T] [J]
        [M] [J] [H] [M] [P] [S] [V] [L] [N]
        [P] [C] [N] [T] [S] [F] [R] [G] [Q]
        [Z] [P] [S] [F] [F] [T] [N] [P] [W]

Into something else based on text inputs"""

import copy 
def load_data():
    with open("data/05_giant_cargo_crane.txt", "r") as f:
        return f.readlines()

def load_crates(txt):
    "Fills 9 arrays based on initial txt file"
    cargo = [[] for _ in range(9)]

    for text in txt[0:8]:
        for i in range(0, len(text), 4):
            if text[i] == '[':
                array_index = i // 4
                cargo[array_index].append(f"{text[i]}{text[i+1]}{text[i+2]}")

    return cargo

def extract_coordinates(text_input):
    """Extracts the coordinates from text input"""
    return [[int(i) for i in line.split() if i.isnumeric()] for line in text_input]

def rearrange_crates(crates, coordinates, reverse):
    """Rearrange the crates so that only one is moved at a time
        meaning that the crate on top will go to the bottom and vice versa"""

    start = 0
    for command in coordinates:
        amount, get_from, put_on = command[0]

        crates_to_move = crates[get_from-1][0:amount]

        if reverse:
            crates_to_move.reverse()

        for crate in crates_to_move[start:]:
            #Appends crates to their new array
            crates[put_on-1].insert(0, crate)

            #Removes crates from orginal array
            crates[get_from-1].pop(0)

    for i in crates:
        print(i)
    print()

file = load_data()
cargo = load_crates(file)
coordinates = extract_coordinates(file[10:])

crates = copy.deepcopy(cargo)
rearrange_crates(crates, coordinates, reverse = False)
rearrange_crates(cargo, coordinates, reverse = True)

File: Advent_of_code/Advent_of_code_05.cpp
#include <iostream> 
#include <string>
#include <fstream>
#include <vector>
#include <sstream>
#include <algorithm>

using namespace std;

/* For this task we will rearrange some boxes that look like this:

                    [L] [M]         [M]    
                [D] [R] [Z]         [C] [L]
                [C] [S] [T] [G]     [V] [M]
        [R]     [L] [Q] [B] [B]     [D] [F]
        [H] [B] [G] [D] [Q] [Z]     [T] [J]
        [M] [J] [H] [M] [P] [S] [V] [L] [N]
        [P] [C] [N] [T] [S] [F] [R] [G] [Q]
        [Z] [P] [S] [F] [F] [T] [N] [P] [W]

Into something else based on text inputs */

using row = vector<string>;
using matrix = vector<vector<char>>;

row get_data(const string& filename) {
    row data;
    ifstream file(filename);
    string line;

    if (file.is_open()) {
        // Read each line and add it to the vector
        while (getline(file, line)) {
            if (line.empty()) continue;

            data.push_back(line);
        }
        file.close();
    }
    return data;
}

matrix load_crates(row data){
    int array_index = 0;
    int row_with_crates = 8;
    matrix crates = {{}, {}, {}, {}, {}, {}, {}, {}, {}};

    for (int i = 0; i < row_with_crates; i++){
        for (int j = 1; j <= data[i].size(); j+=4) {
            if (data[i][j] != ' '){
                array_index = (int) j / 4;
                crates[array_index].push_back(data[i][j]);
            }
        }
    }
    return crates;
}

vector<int> load_coordinates(row data){
    vector<int> commands;
    string word;
    
    for(int i=9; i < data.size(); i++){
        istringstream iss(data[i]);

        while (iss >> word){            
            if (isdigit(word[0])) {
                commands.push_back(stoi(word));
            }; 
        }
    }
    return commands;
}

int rearrange_crates(matrix crates, vector<int> coord, bool inverse=false){

    int amount, origin, target;

    for (int i = 0; i < coord.size() -2; i += 3){
        // -1 is used due to array indexes statrting at 0
        amount = coord[i];
        origin = coord[i+1] -1;
        target = coord[i+2] -1;

        // Loads correspond crates and inverts the array if necessary
        vector<char> moving_crates(crates[origin].begin(), crates[origin].begin() + amount);
        if (inverse == true) reverse(moving_crates.begin(), moving_crates.end());

        for (int j = 0; j < moving_crates.size(); j++){           
            // Appends crates to their new array
            crates[target].insert(crates[target].begin(),  moving_crates[j]);

            // Removes crates from orginal array
            crates[origin].erase(crates[origin].begin());
        }
    }
    // Prints the final arrangement from crates
    for (int k = 0; k < crates.size(); k++){
    for (int l = 0; l < crates[k].size(); l++){
                cout << crates[k][l] << ' ';
        }cout << endl;
    }
    cout << endl;
    return 0;
}

int main() {
    row data = get_data("data/05_giant_cargo_crane.txt");
    matrix crates = load_crates(data);
    vector<int> coordinates = load_coordinates(data);

    rearrange_crates(crates, coordinates);
    cout << "INVERTING "<< endl;
    rearrange_crates(crates, coordinates, true);

    return 0;
}